tools/compiler/src/processor.rs

```diff
@@ -60,21 +60,53 @@ pub async fn generate_registry() -> AppResult<()> {
     Ok(())
 }
 
+/// Given `ibc_data` describing a channel and `source_asset` on the source chain,
+/// compute the metadata for the asset when it is transported along the channel onto a Penumbra chain.
+fn transport_metadata_along_channel(ibc_data: &IbcConfig, source_asset: Metadata) -> Metadata {
+    // The `Metadata` structure doesn't allow modifying the internals, so drop to raw proto data
+    let mut pb_metadata: pb::Metadata = source_asset.into();
+    tracing::debug!(?pb_metadata, "original");
+
+    let prefix_channel = |x: &mut String| {
+        *x = format!("transfer/{}/{}", ibc_data.ibc_channel, x);
+    };
+
+    // Prefix the channel to the base denom.
+    prefix_channel(&mut pb_metadata.base);
+    // Prefix the channel to the display denom.
+    prefix_channel(&mut pb_metadata.display);
+    // Prefix the channel to all denom units.
+    for denom_unit in pb_metadata.denom_units.iter_mut() {
+        prefix_channel(&mut denom_unit.denom);
+    }
+
+    // Delete the asset ID, so that it will be recomputed with the adjusted base denom.
+    // Without this, decoding will fail because the asset ID won't match.
+    pb_metadata.penumbra_asset_id = None;
+
+    tracing::debug!(?pb_metadata, "new");
+    Metadata::try_from(pb_metadata).unwrap()
+}
+
 async fn process_chain_config(chain_config: ChainConfig) -> AppResult<Registry> {
     let mut all_metadata = Vec::new();
     all_metadata.extend(chain_config.native_assets.clone());
 
     // For each ibc connection, fetch all metadata of native assets from the cosmos registry
     let responses = query_github_assets(&chain_config).await?;
 
-    for (ibc_asset, asset_list) in responses {
-        for asset in asset_list.assets {
+    for (ibc_data, asset_list) in responses {
+        for source_asset in asset_list.assets {
             // ICS20 assets should be unwound through their native chains, we can skip
-            if asset.type_asset == AssetTypeAsset::Ics20 {
+            if source_asset.type_asset == AssetTypeAsset::Ics20 {
                 continue;
             }
-            let metadata = convert_to_proto_metadata(&ibc_asset, asset)?;
-            all_metadata.push(metadata);
+            // Turn the asset back into JSON so we can deserialize it as a penumbra Metadata
+            let asset_json = serde_json::to_string(&source_asset)?;
+            let transferred_asset =
+                transport_metadata_along_channel(&ibc_data, serde_json::from_str(&asset_json)?);
+            tracing::info!(?asset_json, transferred_asset_json = ?serde_json::to_string(&transferred_asset));
```

Comment:
Is this outputting something helpful for you? I just see what feels like a bunch of noise. But maybe it's me not familiar with this tracing library.

<img width="1192" alt="Screenshot 2024-04-11 at 1 00 45 PM" src="https://github.com/prax-wallet/registry/assets/16624263/937de43d-8620-4ea7-affa-8fdfb37ba037">

Reply:
Yeah I was using it to inspect the JSON output and determine how the encodings differed during debugging. We can remove